proposal/shared-resources.d.ts.md

```diff
@@ -0,0 +1,247 @@
+  const result2 = await compiler.compileStringAsync('a {b: c}').css;
+  compiler.dispose();
+
+  // throws error
+  const result3 = sass.compileStringAsync('a {b: c}').css;
+}
+```
+
+## API
+
+### Types
+
+```ts
+import {CompileResult} from '../spec/js-api/compile';
+import {Options, StringOptions} from '../spec/js-api/options';
+```
+
+#### initCompiler()
+
+Returns a synchronous [Compiler].
+
+[Compiler]: #compiler
+
+```ts
+export function initCompiler(): Compiler;
+```
+
+#### initAsyncCompiler()
+
+Resolves with an [Async Compiler].
+
+[Async Compiler]: #async-compiler
+
+```ts
+export function initAsyncCompiler(): Promise<AsyncCompiler>;
+```
+
+#### Compiler
+
+An instance of the synchronous [Compiler]. Re-exports [`compile()`]
+and [`compileString()`].
+
+[`compile()`]: ../spec/js-api/compile.d.ts.md#compile
+[`compilestring()`]: ../spec/js-api/compile.d.ts.md#compilestring
+
+```ts
+export interface Compiler {
+  compile(path: string, options?: Options<'sync'>): CompileResult;
+  compileString(source: string, options?: StringOptions<'sync'>): CompileResult;
+```
+
+##### dispose()
+
+Returns the result of [Disposing a Compiler].
+
+[disposing a compiler]: #disposing-a-compiler
+
+```ts
+  dispose(): Boolean;
+}
+```
+
+#### Async Compiler
+
+An instance of the asynchronous [Compiler interface]. Re-exports
+[`compileAsync()`] and [`compileStringAsync()`].
+
+[`compileasync()`]: ../spec/js-api/compile.d.ts.md#compileasync
+[`compilestringasync()`]: ../spec/js-api/compile.d.ts.md#compilestringasync
+
+```ts
+export interface AsyncCompiler {
+  compileAsync(
+    path: string,
+    options?: Options<'async'>
+  ): Promise<CompileResult>;
+  compileStringAsync(
+    source: string,
+    options?: StringOptions<'async'>
+  ): Promise<CompileResult>;
+```
+
+##### dispose()
+
+* Resolves with the result of [Disposing a Compiler].
+
+```ts
+  dispose(): Promise<Boolean>;
+}
+```
+
+## Semantics
+
+### Compiler
+
+A Compiler must:
+
+* Have a lifetime, which starts upon construction.
+
+* `compile` and `compileString` must have identical semantics to the [Sass
+  interface].
+
+* Have a `dispose` method, which ends the Compiler's lifetime.
+
+[sass interface]: ../spec/js-api/index.d.ts.md
+
+### Async Compiler
+
+An Async Compiler must:
+
+* Have a lifetime, which starts upon construction.
+
+* `compileAsync` and `compileStringAsync` must have identical semantics to the
+  [Sass interface].
+
+* Have a `dispose` method, which ends the Compiler's lifetime.
+
+### Disposing a Compiler
+
+When `dispose` is invoked on a Compiler:
+
+* Any subsequent invocations of `compile` and `compileString` must throw an
+  error.
+
+* Returns `true` when disposal is complete.
+
+### Disposing an Async Compiler
+
+When `dispose` is invoked on a Compiler or Async Compiler:
+
+* Any subsequent invocations of `compileAsync` and `compileStringAsync` must
+  throw an error.
+
+* Resolves a Promise with `true` when disposal is complete.
```

Comment:
Why pull these into separate sections instead of just including them directly in the docs for their respective `dispose()` methods?

Reply:
Combined in https://github.com/sass/sass/commit/ff34f2d8e6039df5b0adeeabefdee67edddfbdd5